File: 프로그래머스/2/12936. 줄 서는 방법/줄 서는 방법.java
import java.util.*;

class Solution {
    public int[] solution(int n, long k) {
        Map<Integer, Long> f = new HashMap<>();
        List<Integer> nums = new ArrayList<>();
        
        for (int i = 1; i <= n; i++) {
            nums.add(i);
        }
        k--;
        
        int[] answer = new int[n];
        for(int i=0; i<n; i++) {
            long fact = f.containsKey(n-1-i) ? f.get(n-1-i) : factorial(n-1-i);
            
            int nextIndex = (int) (k/fact);
            answer[i] = nums.remove(nextIndex);
            k %= fact;
        }
        
        return answer;
    }
    
    public long factorial(int n) {
        long result = 1;
        
        for(int i=2; i<=n; i++) {
            result *= i;
        }
        return result;
    }
    
}

File: 프로그래머스/2/12936. 줄 서는 방법/줄 서는 방법.py
import math

def solution(n, k):
    nums =[i for i in range(1, n + 1)]
    answer = []
    fact = math.factorial(len(nums) - 1)
    

    while nums:
        next_index = math.ceil(k / fact) - 1
        answer.append(nums.pop(next_index))
        k -= fact * next_index
        fact //= max(1, len(nums))
    
    return answer

    # k  f  nums     n_i
    # 5  2  [1,2,3]   
    
        
        